utils/compression/zstd_compressor.go

```diff
@@ -0,0 +1,40 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package compression
+
+import (
+	"fmt"
+
+	"github.com/DataDog/zstd"
+)
+
+var _ Compressor = (*zstdCompressor)(nil)
+
+func NewZstdCompressor(maxSize int64) Compressor {
+	return &zstdCompressor{
+		maxSize: maxSize,
+	}
+}
+
+type zstdCompressor struct {
+	maxSize int64
+}
+
+func (z *zstdCompressor) Compress(msg []byte) ([]byte, error) {
+	if int64(len(msg)) > z.maxSize {
+		return nil, fmt.Errorf("%w: (%d) > (%d)", ErrMsgTooLarge, len(msg), z.maxSize)
+	}
+	return zstd.Compress(nil, msg)
+}
+
+func (z *zstdCompressor) Decompress(msg []byte) ([]byte, error) {
+	decompressed, err := zstd.Decompress(nil, msg)
```

Comment:
afaict `zstd.Decompress` does not prevent against zip bombs. ex:

```golang
func TestZstdZipBomb(t *testing.T) {
	compressor := NewZstdCompressor(math.MaxInt64)

	data := make([]byte, units.GiB)
	compressed, err := compressor.Compress(data)
	require.NoError(t, err)

	compressor = NewZstdCompressor(maxMessageSize)

	_, err = compressor.Decompress(compressed)
	require.NotErrorIs(t, err, ErrDecompressedMsgTooLarge)
}
```

This will decompress the full GiB message (over the p2p it's just a few KiB). It does error, but only after doing all the work.